File: writers/16bit/cpuidwrt.pas
{$L CPUID.OBJ}

program CPUIDTP;

uses
  Crt, Dos;

type
  TEntry = record
    Leaf: LongInt;
    SubLeaf: LongInt;
    EAX: LongInt;
    EBX: LongInt;
    ECX: LongInt;
    EDX: LongInt;
  end;

var
  F: Text;

{--------------------------------------------------------------------
  CPUID is declared as an external procedure. It takes two 32‑bit input
  parameters (Leaf and SubLeaf) and four var parameters (to receive the
  32‑bit values returned in EAX, EBX, ECX, and EDX). The routine is
  provided by CPUID.ASM.
--------------------------------------------------------------------}
procedure CPUID(Leaf, SubLeaf: LongInt; var EAX, EBX, ECX, EDX: LongInt); external name '_CPUIDReal';

{--------------------------------------------------------------------
  WriteEntry writes one JSON object (an entry) to the text file.
--------------------------------------------------------------------}
procedure WriteEntry(Leaf, SubLeaf, EAX, EBX, ECX, EDX: LongInt; IsLast: Boolean);
begin
  Write(F, '    { "leaf": ', Leaf, ', "subleaf": ', SubLeaf,
    ', "eax": ', EAX, ', "ebx": ', EBX, ', "ecx": ', ECX, ', "edx": ', EDX, ' }');
  if not IsLast then
    Writeln(F, ',')
  else
    Writeln(F);
end;

var
  maxStandard, maxExtended: LongInt;
  leaf, subleaf: LongInt;
  a, b, c, d: LongInt;
begin
  ClrScr;
  Assign(F, 'cpuid_data.json');
  Rewrite(F);
  Writeln(F, '{');
  Writeln(F, '  "entries": [');

  { --- Capture Standard CPUID Leaves --- }
  { Call CPUID(0,0) which returns the maximum standard leaf in EAX. }
  CPUID(0, 0, maxStandard, b, c, d);
  for leaf := 0 to maxStandard do
  begin
    if (leaf = 4) or (leaf = $B) or (leaf = $D) then
    begin
      subleaf := 0;
      while True do
      begin
        CPUID(leaf, subleaf, a, b, c, d);
        { For leaf 4: if subleaf > 0 and the lower 5 bits of EAX are zero, break. }
        if (leaf = 4) and (subleaf > 0) and ((a and $1F) = 0) then Break;
        { For leaf $B: if subleaf > 0 and EAX is zero, break. }
        if (leaf = $B) and (subleaf > 0) and (a = 0) then Break;
        { For leaf $D: if subleaf > 0 and all registers are zero, break. }
        if (leaf = $D) and (subleaf > 0) and (a = 0) and (b = 0) and (c = 0) and (d = 0) then Break;
        WriteEntry(leaf, subleaf, a, b, c, d, False);
        Inc(subleaf);
      end;
    end
    else
    begin
      CPUID(leaf, 0, a, b, c, d);
      WriteEntry(leaf, 0, a, b, c, d, False);
    end;
  end;

  { --- Capture Extended CPUID Leaves --- }
  CPUID($80000000, 0, maxExtended, b, c, d);
  for leaf := $80000000 to maxExtended do
  begin
    if leaf = $8000001D then
    begin
      subleaf := 0;
      while True do
      begin
        CPUID(leaf, subleaf, a, b, c, d);
        if (subleaf > 0) and ((a and $1F) = 0) then Break;
        WriteEntry(leaf, subleaf, a, b, c, d, False);
        Inc(subleaf);
      end;
    end
    else
    begin
      CPUID(leaf, 0, a, b, c, d);
      WriteEntry(leaf, 0, a, b, c, d, False);
    end;
  end;

  { --- Finalize JSON --- }
  Writeln(F, '  ]');
  Writeln(F, '}');
  Close(F);

  WriteLn('CPUID data captured in cpuid_data.json');
  WriteLn('Press any key to exit...');
  ReadKey;
end.

File: writers/16bit/cpuidwrt.c
#include <stdio.h>
#include <conio.h>

/*
 * We declare the external CPUID routine.
 * The routine is expected to be defined in an external assembly file (CPUID.ASM)
 * and linked into the final DOS executable.
 *
 * Parameters:
 *   leaf, subleaf: 32-bit input values
 *   eax, ebx, ecx, edx: pointers to 32-bit variables that receive the CPUID output.
 */
extern void CPUID(long leaf, long subleaf, long *eax, long *ebx, long *ecx, long *edx);

int main(void) {
    FILE *fp;
    unsigned long maxStandard, maxExtended;
    long eax, ebx, ecx, edx;
    int leaf, subleaf;

    fp = fopen("cpuid_data.json", "w");
    if (fp == NULL) {
        printf("Error opening output file.\n");
        return 1;
    }

    fprintf(fp, "{\n  \"entries\": [\n");

    /* --- Capture Standard CPUID Leaves --- */
    /* CPUID(0,0) returns the maximum standard leaf in EAX */
    CPUID(0, 0, (long *)&maxStandard, &ebx, &ecx, &edx);
    for (leaf = 0; leaf <= maxStandard; leaf++) {
        if ((leaf == 4) || (leaf == 0xB) || (leaf == 0xD)) {
            subleaf = 0;
            while (1) {
                CPUID(leaf, subleaf, &eax, &ebx, &ecx, &edx);
                if ((leaf == 4) && (subleaf > 0) && ((eax & 0x1F) == 0))
                    break;
                if ((leaf == 0xB) && (subleaf > 0) && (eax == 0))
                    break;
                if ((leaf == 0xD) && (subleaf > 0) && (eax == 0) && (ebx == 0) && (ecx == 0) && (edx == 0))
                    break;
                fprintf(fp, "    { \"leaf\": %d, \"subleaf\": %d, \"eax\": %ld, \"ebx\": %ld, \"ecx\": %ld, \"edx\": %ld },\n",
                        leaf, subleaf, eax, ebx, ecx, edx);
                subleaf++;
            }
        } else {
            CPUID(leaf, 0, &eax, &ebx, &ecx, &edx);
            fprintf(fp, "    { \"leaf\": %d, \"subleaf\": 0, \"eax\": %ld, \"ebx\": %ld, \"ecx\": %ld, \"edx\": %ld },\n",
                    leaf, eax, ebx, ecx, edx);
        }
    }

    /* --- Capture Extended CPUID Leaves --- */
    CPUID(0x80000000, 0, (long *)&maxExtended, &ebx, &ecx, &edx);
    for (leaf = 0x80000000; leaf <= maxExtended; leaf++) {
        if (leaf == 0x8000001D) {
            subleaf = 0;
            while (1) {
                CPUID(leaf, subleaf, &eax, &ebx, &ecx, &edx);
                if ((subleaf > 0) && ((eax & 0x1F) == 0))
                    break;
                fprintf(fp, "    { \"leaf\": %d, \"subleaf\": %d, \"eax\": %ld, \"ebx\": %ld, \"ecx\": %ld, \"edx\": %ld },\n",
                        leaf, subleaf, eax, ebx, ecx, edx);
                subleaf++;
            }
        } else {
            CPUID(leaf, 0, &eax, &ebx, &ecx, &edx);
            fprintf(fp, "    { \"leaf\": %d, \"subleaf\": 0, \"eax\": %ld, \"ebx\": %ld, \"ecx\": %ld, \"edx\": %ld },\n",
                    leaf, eax, ebx, ecx, edx);
        }
    }

    /* Finalize JSON (the last trailing comma can be cleaned up manually if needed) */
    fprintf(fp, "  ]\n}\n");
    fclose(fp);

    printf("CPUID data captured in cpuid_data.json\n");
    getch();
    return 0;
}
